### ISSUE TITLE
byte_size doesn't match up with count of mapped blocks

### ISSUE THREAD

**Issue by superr** [cubinator/ext4#2]
Hello again,

I am getting some errors when doing `entry_inode.open_read()` on 2 different raw ext4 Android img files for the same device. [Here is the smaller of the 2 img files](https://www.androidfilehost.com/?fid=890129502657581038). Hopefully the problem is the same for both.

```
line 862, in open_read
    return BlockReader(self.volume, len(self), mapping)
  File "ext4.py", line 959, in __init__
    raise BlockMapError("byte_size doesn't match up with count of mapped blocks")
ext4.BlockMapError: byte_size doesn't match up with count of mapped blocks
```
Most of the inodes read without issue, but the following files in the linked img fail:
```
/bin/cnss_diag
/bin/thermal-engine
/lib/libmmcamera2_sensor_modules.so
/lib/libchromatix_imx179_video_qtr_16_9_3a.so
/lib/libchromatix_imx179_video_16_9_3a.so
/lib/rfsa/adsp/chre_drv_wifi.so
/lib/rfsa/adsp/chre_drv_loc.so
/lib/rfsa/adsp/libnanohub_slpi_skel.so
/lib/libchromatix_imx378_video_hfr_60_3a.so
/lib/libmmcamera_sony_imx179_eeprom.so
/lib/libchromatix_imx378_video_hfr_120_3a.so
/lib/libchromatix_imx378_zsl_snapshot_3a.so
/lib/dsp/EtsiEaacPlusEncAndCmnModule.so.1
/lib/dsp/mmecns_module.so.1
/lib/libchromatix_imx378_video_hfr_240_3a.so
/lib/libmmcamera_sony_imx378_eeprom.so
/lib/libmmcamera_imx179_htc.so
/lib/libchromatix_imx179_zsl_snapshot_3a.so
/lib/libmmcamera_imx378_htc.so
/lib/libchromatix_imx378_video_16_9_3a.so
/lib64/libfastcvopt.so
/firmware/adsp.b05
/firmware/adsp.b02
/firmware/adsp.b03
/firmware/slpi.b03
/firmware/adsp.b08
/firmware/synaptics_bl71.img
/firmware/hotsound.sound_model
/firmware/fpctzappfingerprint.b03
/firmware/slpi.b05
/firmware/slpi.b08
/firmware/widevine.b03
/firmware/nanohub.full.bin
/firmware/cpe_9335.b08
/firmware/slpi.b13
```

Thank you again for all your time. I hope this is not a difficult fix :)

**Comment by cubinator** [comment]
This is not a bug, there is actually something wrong with the image (or at least it doesn't fit the documentation). I take the first file `/bin/cnss_diag` as an example: If you do `inode = volume.root.get_inode("bin", "cnss_diag")` you get inode 200 at offset 0x32380 in the image. Manually looking at the inode's content reveals the following file-to-disk-mapping

    Extent tree @[inode 200]+0x28=0x323A8:
    File blocks 0 to 68 map to disk blocks 0xDF2 to 0xE36
    File blocks 70 to 71 map to disk blocks 0xE37 to 0xE38

which maps every file block except for file block 69, in other words the inode stores 293216 bytes of data (according to the inode's `i_size` field), but maps only 71 blocks or 290816 bytes of data, so there are 2400 bytes missing (not mapped to disk). Using a script I provide at the end of this answer you get the following table

                                                   --- blocks ---
    file                                           mapped  needed  bytes missing
    bin/cnss_diag                                  71      72      2400
    bin/thermal-engine                             630     1239    2491400
    lib/libmmcamera2_sensor_modules.so             91      95      12916
    lib/libchromatix_imx179_video_qtr_16_9_3a.so   16      17      1192
    lib/libchromatix_imx179_video_16_9_3a.so       16      17      1196
    lib/rfsa/adsp/chre_drv_wifi.so                 9       11      4820
    lib/rfsa/adsp/chre_drv_loc.so                  27      29      6952
    lib/rfsa/adsp/libnanohub_slpi_skel.so          23      24      1216
    lib/libchromatix_imx378_video_hfr_60_3a.so     16      17      1196
    lib/libmmcamera_sony_imx179_eeprom.so          7       9       7472
    lib/libchromatix_imx378_video_hfr_120_3a.so    16      17      1192
    lib/libchromatix_imx378_zsl_snapshot_3a.so     16      17      1196
    lib/dsp/EtsiEaacPlusEncAndCmnModule.so.1       50      51      632
    lib/dsp/mmecns_module.so.1                     114     115     3104
    lib/libchromatix_imx378_video_hfr_240_3a.so    16      17      1196
    lib/libmmcamera_sony_imx378_eeprom.so          8       10      7548
    lib/libmmcamera_imx179_htc.so                  15      80      266200
    lib/libchromatix_imx179_zsl_snapshot_3a.so     16      17      1192
    lib/libmmcamera_imx378_htc.so                  21      82      245764
    lib/libchromatix_imx378_video_16_9_3a.so       16      17      1196
    lib64/libfastcvopt.so                          905     913     30912
    firmware/adsp.b05                              555     565     39178
    firmware/adsp.b02                              29      94      263992
    firmware/adsp.b03                              1       26      99552
    firmware/slpi.b03                              17      62      180696
    firmware/adsp.b08                              1541    1570    117320
    firmware/synaptics_bl71.img                    35      50      59744
    firmware/hotsound.sound_model                  5       7       5072
    firmware/fpctzappfingerprint.b03               72      93      85332
    firmware/slpi.b05                              148     169     85904
    firmware/slpi.b08                              12      18      22464
    firmware/widevine.b03                          4       9       18464
    firmware/nanohub.full.bin                      44      45      3892
    firmware/cpe_9335.b08                          0       1       364
    firmware/slpi.b13                              8       40      130492

Here is the script used to create this table. You have to insert the image name in line 71 and I used ascii color codes in the print section at the end. You can remove them if you want.

    from ext4 import *


    def get_extent_tree (inode):
        if (inode.inode.i_flags & ext4_inode.EXT4_EXTENTS_FL) == 0:
            raise Ext4Error("Inline data")

        # Obtain mapping from extents
        mapping = [] # List of MappingEntry instances

        nodes = queue.Queue()
        nodes.put_nowait(inode.offset + ext4_inode.i_block.offset)

        while nodes.qsize() != 0:
            header_offset = nodes.get_nowait()
            header = inode.volume.read_struct(ext4_extent_header, header_offset)

            if not inode.volume.ignore_magic and header.eh_magic != 0xF30A:
                raise MagicError("Invalid magic")

            if header.eh_depth != 0:
                indices = inode.volume.read_struct(ext4_extent_idx * header.eh_entries, header_offset + ctypes.sizeof(ext4_extent_header))
                for idx in indices: nodes.put_nowait(idx.ei_leaf * inode.volume.block_size)
            else:
                extents = inode.volume.read_struct(ext4_extent * header.eh_entries, header_offset + ctypes.sizeof(ext4_extent_header))
                for extent in extents:
                    mapping.append(MappingEntry(extent.ee_block, extent.ee_start, extent.ee_len))

        return mapping


    files = [
        "bin/cnss_diag",
        "bin/thermal-engine",
        "lib/libmmcamera2_sensor_modules.so",
        "lib/libchromatix_imx179_video_qtr_16_9_3a.so",
        "lib/libchromatix_imx179_video_16_9_3a.so",
        "lib/rfsa/adsp/chre_drv_wifi.so",
        "lib/rfsa/adsp/chre_drv_loc.so",
        "lib/rfsa/adsp/libnanohub_slpi_skel.so",
        "lib/libchromatix_imx378_video_hfr_60_3a.so",
        "lib/libmmcamera_sony_imx179_eeprom.so",
        "lib/libchromatix_imx378_video_hfr_120_3a.so",
        "lib/libchromatix_imx378_zsl_snapshot_3a.so",
        "lib/dsp/EtsiEaacPlusEncAndCmnModule.so.1",
        "lib/dsp/mmecns_module.so.1",
        "lib/libchromatix_imx378_video_hfr_240_3a.so",
        "lib/libmmcamera_sony_imx378_eeprom.so",
        "lib/libmmcamera_imx179_htc.so",
        "lib/libchromatix_imx179_zsl_snapshot_3a.so",
        "lib/libmmcamera_imx378_htc.so",
        "lib/libchromatix_imx378_video_16_9_3a.so",
        "lib64/libfastcvopt.so",
        "firmware/adsp.b05",
        "firmware/adsp.b02",
        "firmware/adsp.b03",
        "firmware/slpi.b03",
        "firmware/adsp.b08",
        "firmware/synaptics_bl71.img",
        "firmware/hotsound.sound_model",
        "firmware/fpctzappfingerprint.b03",
        "firmware/slpi.b05",
        "firmware/slpi.b08",
        "firmware/widevine.b03",
        "firmware/nanohub.full.bin",
        "firmware/cpe_9335.b08",
        "firmware/slpi.b13"
    ]


    image = open("vendor.img", "rb") # INSERT FILE NAME OF IMAGE FILE
    v = Volume(image)


    for file in files:
        inode = v.root.get_inode(*file.split("/"))
        mapping = get_extent_tree(inode)

        # Check block count
        actual_block_count = sum(entry.block_count for entry in mapping)
        expected_block_count = (inode.inode.i_size - 1) // v.block_size + 1

        if actual_block_count < expected_block_count:
            bytes_missing = inode.inode.i_size - actual_block_count * v.block_size
            print("\x1b[1;31m{0: <50s}{1: <8d}{2: <8d}{3:d}\x1b[0m".format(file, actual_block_count, expected_block_count, bytes_missing))
        elif actual_block_count > expected_block_count:
            print("\x1b[1;33m{0: <50s}{1: <8d}{2:d}\x1b[0m".format(file, actual_block_count, expected_block_count))
        else:
            print("\x1b[1;32m{0:s} OK\x1b[0m".format(file))

**Comment by superr** [comment]
Thank you for your response :) This sounds like bad news. I am a little confused because mounting the img as a loop device in Linux allows me to copy these files without issue. Do you think there is a way to handle this or are we out of luck?

**Comment by cubinator** [comment]
I don't actually know. I can mount this image and copy these files too. You might wanna check the integrity of those files somehow (by executing software dependent on those libraries or, if possible, compare them with original files somewhere else). Running `fsck.ext4 -fvn vendor.img` didn't reveal any error, but with `debugfs vendor.img` I got this

    debugfs vendor.img
    debugfs 1.43.4 (31-Jan-2017)
    debugfs:  cd bin
    debugfs:  stat cnss_diag

    Inode: 200   Type: regular    Mode:  0755   Flags: 0x80000
    Generation: 0    Version: 0x00000000
    User:     0   Group:  2000   Size: 293216
    File ACL: 69807    Directory ACL: 0
    Links: 1   Blockcount: 576
    Fragment:  Address: 0    Number: 0    Size: 0
    ctime: 0x495c7800 -- Thu Jan  1 09:00:00 2009
    atime: 0x5a5d2177 -- Mon Jan 15 22:47:35 2018
    mtime: 0x495c7800 -- Thu Jan  1 09:00:00 2009
    Extended attributes:
      security.selinux (29) = "u:object_r:cnss_diag_exec:s0\000"
    EXTENTS:
    (0-68):3570-3638, (70-71):3639-3640

which in the last line shows that there actually is no extent mapping the file block 69 to disk. I don't know how copying those files works then, maybe `cp` just inserts zero-bytes into unmapped file blocks.

You could comment out the exception and add a `pass` underneath it, but this probably leads to actual errors in my script, because reading files is highly dependent on those mappings. You could also ignore these files and extract them manually.

**Comment by superr** [comment]
The image I sent is the unmodified factory vendor.img for the Google Pixel. If I use dd to replace the /vendor partition with this vendor.img everything works perfectly.

Unfortunately there is not another way to extract these files manually on Windows. That is part of why I need this actually. That said, I can't ignore the files. I will try a `pass` on the exception and see what happens. I can at least compare the cp file the the new one if it gets written.

Thanks again for your help :)

**Comment by cubinator** [comment]
Well, I can extract those files with my system and provide them via GDrive to you, if that's fine. But I still don't know what happens to those files, when extracting them via `cp`, `pv` or `dd`.

**Comment by superr** [comment]
Thank you for the offer, but I can mount the img and extract the files. However, not everyone using my project is a Linux user as myself. It needs to be cross platform and work on all img files for all devices lol. Sounds like a lot I know, but I have already tested on many other devices and this module works perfectly. It is only this device having trouble so far :)

edit: changed this to my to be sure you know I am not demanding anything from you :)

**Comment by cubinator** [comment]
The core problem is that I don't know, where to get those missing bytes from, when they are not mapped to disk. I can't just throw in zeros or random bytes and hope for the best. And I actually inspected the image manually (in a hex editor and with debugfs) to ensure that my script didn't get those values wrong. I can't fix the script, if the errors it spits out are really errors in the image according to every documentation on ext4 I could find.

**Comment by superr** [comment]
I tried `pass` on the exceptions, and it does write files, but they are not the same as the cp files. 

As I mentioned before, I have [source](https://github.com/nkk71/e2tools/blob/master/sysXtract.c) for a program written in C that extracts ext4 img files. I just used it to extract the vendor.img and the files are identical to the cp files. I still don't know if either of them are valid, but they are both doing the same thing...whatever that might be. Let me know if you want a binary. It depends on [this repo](https://github.com/nkk71/e2fsprogs) as well.

I understand if this is too much to deal with. Thank you for everything :)

EDIT: A little back story, I am currently using the above mentioned binary in my project. The reasons I am looking for a new solution are:
1) The capabilities are not consistent in windows/linux (which means one or both are not correct)
2) When extracting ext4 in Windows, you can't have file.txt and FILE.TXT in the same directory which happens in Linux frequently. I don't know C so I have no control over changing this source to dump these extra files somewhere else.
3) The wonderful guy that wrote it passed away recently.

Thank you again for all your help :) Other than this firmware, your module is working perfectly to solve all the problems.

**Comment by cubinator** [comment]
I have an assumption: I checked the copied file where the file block was not mapped and indeed, it was filled with zeroes. I assume that zero-blocks in a file are optimized out of the file, so they need not be stored on disk. I have to check that though.

**Comment by superr** [comment]
Thank you for looking into it. You are very much appreciated :)

**Comment by cubinator** [comment]
Unfortunately I haven't been able to recreate this behavior. I haven't found any ext4 creation flags, no internal flags I could edit myself, no mount options and writing files with zero-blocks doesn't introduce missing extents. Either your system image actually uses a modified version of ext4, which I can't find any documentation on except for your source code (where I too didn't find any indications on optimizing zero-blocks out of the file) or the image is actually damaged (and those files not needed for the system to operate correctly).

**Comment by superr** [comment]
That is unfortunate. Here is the [latest make_ext4fs source](https://android.googlesource.com/platform/system/extras/+/android-8.1.0_r13/ext4_utils/make_ext4fs.c), which was probably used to create the vendor.img you are working with. Thanks again for looking at this, and I hope this source helps :)

EDIT: changed the source link. Not sure the other one was actually the latest. Sorry about that

**Comment by superr** [comment]
Just updated the source link I sent. Not sure if you are notified when I edit. Thanks again :)

**Comment by superr** [comment]
I just realized that the vendor.img was originally created as sparse, then unsparsed with [simg2img](https://github.com/nkk71/android-simg2img). I am not sure if this matters but I thought it might be worth mentioning.

**Comment by cubinator** [comment]
Sparse just means that not every block group (big chunk) contains a backup of the first block group's volume description (superblock).

**Comment by superr** [comment]
Thanks for the info. I am trying to think of anything that might help figure out how to deal with these missing blocks. 

You said earlier the mount/cp file was just filled with zeroes where the missing block is. Would it be easy to make that happen in your module? I know you may not want this for everyone, but it seems this might be the answer for these types of ext4 img files. Both mount/cp and sysXtract produce identical files when extracting this vendor.img.

**Comment by cubinator** [comment]
I won't implement this in my module since I can't find anything in the documentation and I want to hold my script as close to the documentation as possible, but I crafted a little script that fills in zeroes where there are unmapped file blocks. Here is a little usage exmaple:

    import ext4
    import xtract

    file = open("vendor.img", "rb")
    volume = ext4.Volume(file)

    inode = volume.root.get_inode("bin", "cnss_diag")
    file_bytes = xtract.xtract(inode)

The function `xtract.xtract(inode)` takes an `ext4.Inode` instance `inode` and spits out a `bytes` object containing the file's data. I haven't done many tests (just confirmed that it works for `/bin/cnss_diag`), so if it produces problems, I can try to fix it. Here is the `xtract` module

    from ext4 import *

    def xtract (inode):
        if (inode.inode.i_flags & ext4_inode.EXT4_EXTENTS_FL) != 0:
            # Obtain mapping from extents
            mapping = [] # List of MappingEntry instances

            nodes = queue.Queue()
            nodes.put_nowait(inode.offset + ext4_inode.i_block.offset)

            while nodes.qsize() != 0:
                header_offset = nodes.get_nowait()
                header = inode.volume.read_struct(ext4_extent_header, header_offset)

                if not inode.volume.ignore_magic and header.eh_magic != 0xF30A:
                    raise MagicError(f"Invalid magic value in extent header at offset 0x{header_offset:X} of inode {inode.inode_idx:d}: 0x{header.eh_magic:04X} (expected 0xF30A)")

                if header.eh_depth != 0:
                    indices = inode.volume.read_struct(ext4_extent_idx * header.eh_entries, header_offset + ctypes.sizeof(ext4_extent_header))
                    for idx in indices: nodes.put_nowait(idx.ei_leaf * inode.volume.block_size)
                else:
                    extents = inode.volume.read_struct(ext4_extent * header.eh_entries, header_offset + ctypes.sizeof(ext4_extent_header))
                    for extent in extents:
                        mapping.append(MappingEntry(extent.ee_block, extent.ee_start, extent.ee_len))

            # Optimize mapping for faster read
            mapping.sort(key = lambda entry: entry.file_block_idx)

            idx = 0
            while idx < len(mapping):
                while idx + 1 < len(mapping):
                    if mapping[idx].file_block_idx + mapping[idx].block_count > mapping[idx + 1].file_block_idx:
                        raise Ext4Error("Same file block mapped to multiple disk blocks!")
                    elif mapping[idx].file_block_idx + mapping[idx].block_count == mapping[idx + 1].file_block_idx:
                        tmp = mapping.pop(idx + 1)
                        mapping[idx].block_count += tmp.block_count
                    else:
                        break

                idx += 1

            # Read data
            chunks = []
            for idx, entry in enumerate(mapping):
                # Read mapped blocks
                chunks.append(inode.volume.read(entry.disk_block_idx * inode.volume.block_size, entry.block_count * inode.volume.block_size))

                # Check for unmapped file blocks
                if idx + 1 < len(mapping) and entry.file_block_idx + entry.block_count < mapping[idx + 1].file_block_idx:
                    blocks_missing = mapping[idx + 1].file_block_idx - (entry.file_block_idx + entry.block_count)
                    chunks.append((b"\0" * inode.volume.block_size) * blocks_missing)

            # Check for missing bytes at the end
            bytes_read = sum(len(chunk) for chunk in chunks)
            if bytes_read >= inode.inode.i_size:
                return b"".join(chunks)[:inode.inode.i_size] # TODO Can be solved without as much copying
            else:
                bytes_missing = inode.inode.i_size - bytes_read
                chunks.append(b"" * bytes_missing) # TODO Can be slow if bytes_missing is huge
                return b"".join(chunks)
        else:
            # Inode uses inline data
            i_block = inode.volume.read(inode.offset + ext4_inode.i_block.offset, ext4_inode.i_block.size)
            return i_block

**Comment by superr** [comment]
You are a genius !!! Thank you so much. The files are now exactly the same as the mount/cp and sysXtract files :)

I assume it is safe to run all inodes through `xtract.xtract(inode)`, is this correct? For example, will it cause problems for the system.img I sent before or should I only use this when I am working on this kind of img?

**Comment by cubinator** [comment]
The script ALWAYS reads SOMETHING, even if the inode isn't even a file or in-use at all, because it just checks whether or not the file is mapped via extents in which case it reads the mapped blocks, but otherwise it reads the inode's inline data regardless whether it actually contains data (it wouldn't cause any errors though).

So use it only on inodes you got from directories, because then you can be sure that this inode actually is in-use and contains data.

(When you are still recursively following the root directory, you should be fine with every file you encounter.)

**Comment by superr** [comment]
Thank you for all your time. It seems to work perfectly :)

**Comment by superr** [comment]
Hello,

After more testing with the xtract module, some of the files that are missing blocks are not identical to the mounted img files. Below is a table created from the script you posted before, including all the files that are not identical after running through the xtract module.
```
system/app/Chrome/Chrome.apk                      31966   31979   52745
system/app/GoogleCamera/oat/arm64/GoogleCamera.vdex1474    1479    17847
system/app/LatinIMEGooglePrebuilt/LatinIMEGooglePrebuilt.apk9472    9499    109955
system/fonts/NotoColorEmoji.ttf                   1776    1778    4876
system/framework/oat/arm64/wifi-service.vdex      442     448     23579
system/lib64/libicuuc.so                          397     398     8
system/lib64/libradio.so                          21      22      8
system/priv-app/NexusLauncherPrebuilt/oat/arm64/NexusLauncherPrebuilt.vdex549     552     10272
system/priv-app/SettingsGoogle/oat/arm64/SettingsGoogle.vdex1500    1503    9882
system/priv-app/SystemUIGoogle/oat/arm64/SystemUIGoogle.vdex2118    2125    25967
system/priv-app/Velvet/Velvet.apk                 11347   11349   5925
```

Here is a link to the [system.img](http://sr-code.com/cubinator/filemanager.php?file=system.img_test.zip) that includes the above files.

Additionally, some files that do not have missing blocks do not work properly when run through the xtract module. I have solved this by trying the normal way first, and only running through xtract when an exception is raised.

Thank you so much for everything :)

**Comment by cubinator** [comment]
Does this extraction script work better?

    from ext4 import *

    def xtract (inode):
        if (inode.inode.i_flags & ext4_inode.EXT4_EXTENTS_FL) != 0:
            # Obtain mapping from extents
            mapping = [] # List of MappingEntry instances

            nodes = queue.Queue()
            nodes.put_nowait(inode.offset + ext4_inode.i_block.offset)

            while nodes.qsize() != 0:
                header_offset = nodes.get_nowait()
                header = inode.volume.read_struct(ext4_extent_header, header_offset)

                if not inode.volume.ignore_magic and header.eh_magic != 0xF30A:
                    raise MagicError(f"Invalid magic value in extent header at offset 0x{header_offset:X} of inode {inode.inode_idx:d}: 0x{header.eh_magic:04X} (expected 0xF30A)")

                if header.eh_depth != 0:
                    indices = inode.volume.read_struct(ext4_extent_idx * header.eh_entries, header_offset + ctypes.sizeof(ext4_extent_header))
                    for idx in indices: nodes.put_nowait(idx.ei_leaf * inode.volume.block_size)
                else:
                    extents = inode.volume.read_struct(ext4_extent * header.eh_entries, header_offset + ctypes.sizeof(ext4_extent_header))
                    for extent in extents:
                        mapping.append(MappingEntry(extent.ee_block, extent.ee_start, extent.ee_len))

            # Optimize mapping for faster read
            mapping.sort(key = lambda entry: entry.file_block_idx)

            idx = 0
            while idx < len(mapping):
                while idx + 1 < len(mapping):
                    if mapping[idx].file_block_idx + mapping[idx].block_count > mapping[idx + 1].file_block_idx:
                        raise Ext4Error("Same file block mapped to multiple disk blocks!")
                    elif mapping[idx].file_block_idx + mapping[idx].block_count == mapping[idx + 1].file_block_idx \
                            and mapping[idx].disk_block_idx + mapping[idx].block_count == mapping[idx + 1].disk_block_idx:
                        tmp = mapping.pop(idx + 1)
                        mapping[idx].block_count += tmp.block_count
                    else:
                        break

                idx += 1

            # Read data
            chunks = []
            for idx, entry in enumerate(mapping):
                # Read mapped blocks
                chunks.append(inode.volume.read(entry.disk_block_idx * inode.volume.block_size, entry.block_count * inode.volume.block_size))

                # Check for unmapped file blocks
                if idx + 1 < len(mapping) and entry.file_block_idx + entry.block_count < mapping[idx + 1].file_block_idx:
                    blocks_missing = mapping[idx + 1].file_block_idx - (entry.file_block_idx + entry.block_count)
                    chunks.append((b"\0" * inode.volume.block_size) * blocks_missing)

            # Check for missing bytes at the end
            bytes_read = sum(len(chunk) for chunk in chunks)
            if bytes_read >= inode.inode.i_size:
                return b"".join(chunks)[:inode.inode.i_size] # TODO Can be solved without as much copying
            else:
                bytes_missing = inode.inode.i_size - bytes_read
                chunks.append(b"" * bytes_missing) # TODO Can be slow if bytes_missing is huge
                return b"".join(chunks)
        else:
            # Inode uses inline data
            i_block = inode.volume.read(inode.offset + ext4_inode.i_block.offset, ext4_inode.i_block.size)
            return i_block

**Comment by superr** [comment]
Yes, thank you! Unfortunately some of the same files are still not identical to the mounted system.img files:
```
system/fonts/NotoColorEmoji.ttf
system/framework/oat/arm64/wifi-service.vdex
system/lib64/libicuuc.so
system/lib64/libradio.so
system/priv-app/NexusLauncherPrebuilt/oat/arm64/NexusLauncherPrebuilt.vdex
system/priv-app/SystemUIGoogle/oat/arm64/SystemUIGoogle.vdex
```
[Here is a zip](http://sr-code.com/cubinator/filemanager.php?file=file_compare_fixed.zip) containing the python extracted files, and the cp mounted files that are inconsistent.

It is a step in the right direction :) Thanks again for your time and help :)

EDIT: changed the link mentioned below

**Comment by superr** [comment]
Sorry, I added 2 wrong files to the zip. [This zip](http://sr-code.com/cubinator/filemanager.php?file=file_compare_fixed.zip) has all the correct files.

Well, I added an extra file `SystemUIGoogle.apk` to the mount_files. Please ignore it as this file is identical. There are only 6 inconsistent files now and they are in the extract_files directory. Thanks again and sorry for the mix-up :)

**Comment by cubinator** [comment]
Ok, I uploaded a new version of the module where unmapped blocks are no longer treated as errors, but are read as empty blocks (basically implemented the idea behind the `xtract` script into the `ext4`module). I batch-compared the output of the new `inode.open_read().read()` output with the `cp`ed files of the mounted image and they were identical. Let me know whether it works for you as well.

I didn't find any documentation on empty blocks not being mapped for optimization, however I found many posts online regarding "sparse files" (which are not part of the documentation) only mapping non-empty file-blocks to disk. That's reason enough for me to implement it :)
